### ISSUE TITLE
Adding ngrx-forms implementation into platform

### ISSUE THREAD

**Issue by tashoecraft** [ngrx/platform#431]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
</code></pre>

## What is the current behavior?

There is no ngrx-forms
## Expected behavior:

Either bring in https://github.com/MrWolfZ/ngrx-forms into the platform, our build something that accomplishes the same goal. 

## Other information:
Previous discussion: https://github.com/ngrx/core/issues/12

**Comment by fxck** [comment]
Here's what I messaged to Brandon couple of weeks ago.

![image](https://user-images.githubusercontent.com/1303561/31217424-36b1ad98-a9b7-11e7-8d04-e6d80677246f.png)

**Comment by fxck** [comment]
@brandonroberts @MikeRyanDev @robwormald @vsavkin 

_Any_ update would be really lovely. The last thing you could call an official, public, update on this was [in January](https://github.com/ngrx/core/issues/12#issuecomment-273672301). 

Even "we are currently not planing on doing this because _n_" or "we do not plan on doing this ourselves in foreseeable future and we do not trust people outside ngrx team enough to support any official library, so the best option is you try do it third party" would be really really great. 

Don't take the [angular approach](https://github.com/angular/angular/issues/13248#issuecomment-296907679) of never addressing things.. please.

**Comment by robwormald** [comment]
@fxck I’m not sure the value add here over angular’s forms reactive forms implementation. In fact, I’m not even sure that ngrx should have *anything* to do with forms - which are almost always best treated as local component state. 

It’s not clear *what* you’re proposing, why, or what the benefits and drawbacks are. What problems are you solving? 

 Present a coherent proposal, and we’ll discuss it. “Take over this other repo from a third party” does not constitute such a proposal.

**Comment by fxck** [comment]
@robwormald this issue https://github.com/ngrx/core/issues/12 is full of reasons and given its got almost 60 likes, I'd say I'm not the only one interested in ngrx forms (and honestly, almost 8k stars on [redux forms](https://github.com/erikras/redux-form) speaks for itself as well)

> which are almost always best treated as local component state.

I beg to differ, from my experience, if I had form state in store, my life would be million times easier, being able to reset form state upon success action / effect is one of many use cases

> Present a coherent proposal, and we’ll discuss it. “Take over this other repo from a third party” does not constitute such a proposal.

This I completely agree with https://github.com/MrWolfZ/ngrx-forms is only a bridge between angular forms and ngrx and not the best one at that, what I suggest (read my message to Brandon) is completely standalone module.. and in fact, @MikeRyanDev was already working on a design doc, here are couple of snippets from it:


### Async Validators

```typescript
@Injectable()
class OnBlurNameUniqueValidator {
  constructor(
    private formValidator: FormValidator,
    private nameService: NameService
  ) { }
  
  @Effect() validate$ = this.formValidator
    .createAsyncValidator('isNameUnique')
    .mergeMap(({ formControl, createValidationAction }) =>
      formControl.blur$
        .withLatestFrom(formControl.value$, (_, name) => name)
        .switchMap(name =>
          this.nameService.isUnique(name)
            .map(isUnique => createValidationAction({
              valid: isUnique
            }))
            .catch(() => Observable.of(
              createValidationAction({ valid: false })
            ))
        )
    );
}
```

### custom controls

```typescript
interface FormControl<T> {
  disabled$: Observable<boolean>;
  value$: Observable<T>;
  
  acceptValueChanges(value$: Observable<T>): void;
  acceptTouchedChanges(value$: Observable<void>): void;
}

@Directive({
  selector: 'input',
  host: {
    '(input)': 'valueChanges.next($event.target.value)',
    '(blur)': 'touchedChanges.next()',
    '[disabled]': 'disabled',
    '[value]': 'value'
  }
})
export class InputFormControl {
  valueChanges = new Subject<string>();
  touchedChanges = new Subject<void>();
  disabled: boolean = false;
  value: string = '';

  constructor(control: FormControl<string>) {
    control.disabled$.subscribe(disabled => this.disabled = disabled);
    control.value$.subscribe(value => this.value = value);
    control.acceptValueChanges(this.valueChanges);
    control.acceptTouchedChanges(this.touchedChanges);
  }
}
```

### creating forms 

```

import { group, control } from '@ngrx/forms';

@Component({
  template: `
    <form ngrxForm="nameAndAge">
      <label>
        Name
        <input type="text" ngrxFormControl="name">
      </label>
      
      <label>
        Age
        <input type="number" ngrxFormControl="age">
      </label>
    </form>
  `
})
export class SomeFormComponent {
  form = group({
    name: control({
      disabled: false,
      asyncValidators: [ 'isNameUnique' ]
    }),
    age: control({
      disabled: false,
      validators: [ isOldEnough ],
    })
  });
}
 
function isOldEnough(age: number) {
  return {
    isOldEnough: {
      valid: age > 21
    }
  };
}
```

**Comment by tashoecraft** [comment]
@robwormald Wasn't really thinking that ngrx would pull in some random 3rd party library, more so to see if there were thoughts into building one and incorporating it into the platform as the conversation hasn't progressed in almost a year. 

I'm sure how most people use forms it wouldn't be beneficial to incorporate it into the store, but when the forms get large and complicated, redux really seems to help. My app can have 50+ fields that can change at any point depending on seed data (so dynamic forms), with some complex validations that modify the form in many ways (fields hiding and showing, required if x field has value y). I've gotten around the limitations, but I feel that a central store for all the form data would make it all more robust and easier to work with.

I know a huge amount of work has gone into angular/forms, and it would be a massive undertaking to try and replicate a similar feature set. I've looked into extending the existing forms library to try and incorporate into ngrx, but the time to do it hasn't really been available. I'm mostly curious to see what the community thinks as a solution, and if that means a new module in the platform, a 3rd party library, or merely docs on how to go about it with the current setup then I'm all helping out (as I'm sure many others are as well).

**Comment by fxck** [comment]
So is this back in discussion?

**Comment by fxck** [comment]
Another place where ngrx/forms would be useful are guards, when you want to use `canDeactivate` depending on some form state.

**Comment by kosz** [comment]
One of the core principles of Redux is single source of truth, and without such a connector, the source of truth becomes duplicated in an Angular specific class instance that lives inside a component and now encapsulates state information that the central store can't see without specific implementation work from the developers

**Comment by fxck** [comment]
> but the latest version lives inside the angular reactive form.

Yeah, that's a classic.

```
  currentText$ = this.post$
    .filter((post) => post)
    .take(1)
    .switchMap(
      (post) => this.contentForm.instance
        .get('text')
        .valueChanges
        .startWith(post.text)
    );
```

**Comment by fxck** [comment]
Another problem is, that you don't always have access to the form instance, I usually try to put  the form instance into the "smart" component and pass it down to the dumb components, but there are cases where you want to show current data in a component but the form instance is in different component tree branch altogether and if this happens to be a component that is a route, you can't even put the form instance to the common component tree parent node and pass it down with inputs. So you end up saving the form instance into a service.. and basically replicating store. It sucks.

**Comment by fxck** [comment]
Also optimistic updates. You send request, add data to store, reset form.. it fails, what now? Again, you need to save form state before the request happens somewhere, inject effects to get to the fail stream (to restore data), or success stream (to reset form). Words can't describe how a proper forms integration would help me in daily work.

**Comment by Kaffiend** [comment]
> Words can't describe how a proper forms integration would help me in daily work.

I couldn't agree more. More concise, granular control over complex forms via a proper integration would be a godsend.

**Comment by tashoecraft** [comment]
@robwormald Sorry for pinging you back here as I know you are busy (working on Angular Elements which I’m super excited for). I just wanted to try and get some more info here as we are around a year since the original issue was filled. 

So what I didn’t want this issue to come across as is “I would like this big, large new feature, go out and build it” as I know this is a open source project not backed by anyone that is very stable. 

@fxck  I feel has been pretty vocal about the reasons why we think this new addition would make sense. If you disagree with what’s posted I would love to here your counterpoints to it and maybe we can get work on some docs on how to approach some of these situations. Or I can work on a sample project that highlights some of the issues the community and myself have found, and perhaps it’s a lack of knowledge that that could be worked around. 

I am more then happy to get my hands dirty and work on something that could work, but I don’t feel experienced enough to architect out this solution without guidance from some of the top contributors on here (and have it ultimately help no one but myself and not be adopted).

If perhaps @brandonroberts , @MikeRyanDev could also weigh in on why or why not this should be built, then we could hopefully move forward on something. Whether that is building something into the platform, are even just mentioning forms in the docs in some way, I think it would help those who come to ngrx looking for how they handle their forms.

**Comment by fxck** [comment]
There was some activity on Mike's old design doc (the one I pasted those snippets from [here](https://github.com/ngrx/platform/issues/431#issuecomment-336033756)) over the holidays. So hopefully we'll get some official reply on what's going on, what the plans are, eventually.

**Comment by fxck** [comment]
To anyone interested on the "official" update, you might want to read [this gitter discussion](https://gitter.im/ngrx/platform?at=5a5d3990b48e8c3566e44f01).

Anyway, I've need thinking about how to deal with some of the problems mentioned both here and in https://github.com/ngrx/core/issues/12, mainly these two use cases:

1) how to initially fill in the data
2) how to reset form state (either fully, including values, or [setting it to pristine](https://stackoverflow.com/questions/43217078/angular-redux-ngrx-state-update-vs-forms#comment73506985_43217078) and resetting `submitted` state.) on any success (add, update)

I decided to go with a service, rather than saving a big mutating class (`FormGroup`) to actual store, which seemed like a bad idea. 

The idea was to:

- create all the form instances (`FormGroup`/`FormControl`) inside the topmost module / route they belong to
- "register" them inside my service, that was provided on the module, under a unique id
- inject the service in a smart component, get form instance, pass it down to a dumb form component that has the actual form elements 
- emit an event after this dumb component was `AfterViewInit`ed, passing up reference to `NgForm` and updating the service with reference to it
- inject the service into the effect and use it 
  - inside a `tap` of load success to `patchValue` 
  - inside update success to reset value
  - inside update request to get the current values

There are couple of problems.

1) it's a lot of code, especially getting the `NgForm` reference to the service
2) it's impossible (or at least I haven't figure out) to get the current observable value, since you can't use `withLatestFrom(this.formService.get('key').instance.valueChanges)` because `get` would be called immediately, at time the instance wouldn't yet be registered and you can't use `switchMap(() => this.formService.get('key').instance.valueChanges)` either, because it would be empty until the next change, so you have to resort to using sync value `map(() => this.formService.get('key').instance.value)` 

I ultimately decided it's way too much code, way too much boilerplate, way too hacky solution. Angular (not so) reactive forms simply do not fit well into ngrx ecosystem.

The way I see it, given that there won't really be an official `@ngrx/forms` package we, as a community, or the part interested in forms, have two options.

Embrace https://github.com/MrWolfZ/ngrx-forms give it enough feedback to make it a thing (which I'm not a fan of, it being only a bridge between angular forms will bite us in the arse eventually), or try and create coordinated effort to create our own `@ngrx/forms`, perhaps taking as much ideas from popular matured redux forms libraries as possible..

**Comment by Kaffiend** [comment]
@fxck Maybe one of us should startup a repo for a lib, i would contribute, as i'm desperate for a better solution to my plethora of forms in my applications, both professionally (my staff included) and personally. I too am not a fan of `ngrx-forms`.

**Comment by MrWolfZ** [comment]
@fxck @Kaffiend could you elaborate on why you are not fans of `ngrx-forms`? I know I have made some opinionated design decisions and I would be really interested in whether it would be possible to change the library to fit your requirements.

In terms of it being just a bridge between `@angular/forms` and ngrx I cannot follow that argument. The only interaction/dependency is the re-use of the `ControlValueAccessor` concept to allow easier integration with third party components that were designed for `@angular/forms` (which there are many of that I am sure wouldn't want to make ngrx-specific changes).

**Comment by Kaffiend** [comment]
@MrWolfZ [this](https://github.com/MrWolfZ/ngrx-forms/blob/master/example-app/src/app/simple-form/simple-form.component.ts#L17) is the main thing that turned me off from `ngrx-forms` There may be other ways to utilize the library, and perhaps you can enlighten me :) but i have no interest in implementing forms in that manner. Beyond that i have not looked further into the library. I would just as well add another reducer than template it in a string. Telerik took a similar approach to templates in Kendo UI and it was always more pain than gain, until they decided to rebuild native.

**Comment by MrWolfZ** [comment]
@Kaffiend I am not sure I get what you mean. The code you linked is from the example application and is only for outputting the example code in the UI of the example application itself (if you browse to the example application [here](https://ngrx-forms-example-app-v2.herokuapp.com/simpleForm) you can see that there are tabs at the top that show the code). What you linked is in no way being executed or required to use ngrx-forms. I simply copy & pasted the code into the component. Maybe I should clarify this with a comment in the code.

If you look at the other files besides the component [here](https://github.com/MrWolfZ/ngrx-forms/tree/master/example-app/src/app/simple-form) you can see that the code in the code files is (almost) the same as the one inside the component.

But maybe I am simply misunderstanding your post?

Also, if it would be desired to create actual form element components (like redux forms does, not sure if you mean that by "native") I think that would be easy enough to achieve as a layer on top of ngrx-forms.

**Comment by Kaffiend** [comment]
I think i'm misunderstanding the example and the implementation. Apologies, i need to take another look after some coffee 😄

**Comment by Kaffiend** [comment]
So, @MrWolfZ it appears I greatly underestimated the library and I make no excuse as to why, plain and simple i skimmed the library and formed an assumption and it was wrong. After finding the documentation and getting started, that i clearly missed the first time, all was apparent. Seeing code in comments or strings haunts me from the horror show of Kendo UI templates in its early days for form integrations.

**Comment by MrWolfZ** [comment]
Sorry for making you relive the nightmares from the past ;)

If you have any suggestions on how I can improve the library I'd be happy to hear them (and accept contributions of course). ngrx-forms is my first open source project and its interesting to see how the ecosystem works and how libraries evolve.

**Comment by fxck** [comment]
@MrWolfZ it's been a while since I looked into it, but what I'm sure that two things I disliked the most were the idea of having to manually put each form into the state and do your own reducing. This should be the responsibility of the form. The other thing was using standard select to get the form state and then using that to "set control state".

The way I imagine this should work is:

- you put `FormsConnectingModule` into you app module, which internally creates a feature state called forms (or it can be optional, just as `StoreRouterConnectingModule.forRoot({ stateKey: 'router' })` is)
- in your component, you wouldn't have to use store to get the controls, you'd just use a path, either

```html
<form ngrxFormState="uniqueKey">
  <input type="text" ngrxFormControlState="someTextInput">
</form>
```

similar to what you can do with `formControlName`, alternatively, you'd provide the full path `uniqueKey.someTextInput` and it would internally select `stateKey.uniqueKey.someTextInput`.

Some other things I'd consider:

1) tracking more state than what angular forms do, such as `focus` and `blur` 
2) providing selectors / a wrapper service around `Store`...
```typescript
constructor(private _forms: NgrxForms) {
  value$ = this._forms.pipe(
    byKey<FormKeyState>('uniqueKey'), 
    value() // errors(), focused(), valid().. 
  );
}
```

**Comment by fxck** [comment]
Look at @MikeRyanDev snippets I posted here https://github.com/ngrx/platform/issues/431#issuecomment-336033756 I think that should answers first two points ("these would fire register form / field action, creating the forms state shape.." was added in an edit, I'm all for creating the controls beforehand, while still keeping the `path` approach and thus being able to validate it).

**Comment by fxck** [comment]
What do you think, @MrWolfZ ?

**Comment by fxck** [comment]
I imagine it could work like this.

- you add `NgrxFormStoreConnectingModule.forRoot({ stateKey: 'forms' })` to your app module, which creates a feature state `forms` with `StoreModule.forFeature(stateKey, ngrxFormsReducer)` and provides `NgrxFormStore`
- you inject `NgrxFormStore<FormsState>` into the module of the module you you have your form in, you call `NgrxFormStore.register('uniqueFormKey', NgrxFormsGroup)`, where NgrxFormsGroup would be something along the lines of Mike's snippet, this NgrxFormsGroup, this register would dispatch `[ngrx-forms] Register Control` with the structure defined with the `NgrxFormsGroup` helper
- your directives then accepts dot path to the form state, which you can validate whether they exist
- if you need access to form store state, you'd call `NgrxFormStore.state$.pipe(select('dot.path'))`, the reason you'd use `NgrxFormStore.state$` instead of `Store` is because NgrxFormStore has internal knowledge of `stateKey` defined in root

@MrWolfZ

**Comment by MrWolfZ** [comment]
@fxck sorry, I didn't have any time during the last couple of days due to work stress. I'll answer you properly as soon as possible, but potentially it'll have to wait until the weekend :(

**Comment by MrWolfZ** [comment]
@fxck So, I finally found some time to write up my thoughts.

First of all, thank you for taking the time to come up with these ideas. However, I am missing a crucial part in your argument. You have a lot of ideas of how the API could look like. However, what I am even more interested in is _why_ the API should look like that. What are the benefits, what are the drawbacks? Below I will give you the pros and cons I see in your suggestions. And of course I will also give you my "_why_" for the API design I chose for ngrx-forms.

Pros:

- less boilerplate
- easier to debug due to all interaction with the form state happening through explicit actions (thereby being captured by tools like the redux devtools)
- defining the form structure and sync validation functions in the component might be more intuitive for users
- creating async validators as re-usable effects is a very cool idea that can save quite a bit of boilerplate

Cons:

- no compile-time validation of form structure/control access
- having most logic in components or services seems anti-redux
- by only having the form state structure be defined implicitly through directives in the template or a form instance in the component instead of through an explicit state interface you always have to access controls by their name as a string instead of having the type system (and thereby IDE autocomplete) tell you/validate what controls exist on a given form (e.g. having to use `NgrxFormStore.state$.pipe(select('form.someControl'))` instead of just `store.select(s => s.path.to.form.controls.someControl)`)
- the form control directives that only take the name and find their parent group/array to register themselves have some issues with partial forms as discussed in [this issue](https://github.com/angular/angular/issues/10251)
- dependency on rxjs (not per se bad since Angular apps will reference rx anyways but the less own dependencies the library has the better)

Now to my "_why_" for the design I chose:

I love functional programming. I also love reactive programming (and therefore all rx libraries). I have used both in many projects. However, in comparison of the two functional programming will always win for me. Therefore I decided that a) I want the full API to be properly and statically typed and b) have almost all the logic in pure functions and reducers. Also, by designing the API to be best accessible through reducers users are encouraged to put their own heavy-lifting logic in reducers as well (something I have often seen done wrong). The library and any user code also become a breeze to test due to all logic being inside reducers. I would pay for these benefits with some boilerplate any day.

As I mentioned before, I am always willing to extend the API to fit more use cases. However, since I would not be using the API in the ways you suggest (since the current design works just fine for everything I use it for) it would be difficult for me to properly iterate upon it. I strongly believe that the best APIs are designed by people that use them themselves (or as the saying goes "eat your own dogfood"). Therefore it would be great if you (or someone else who likes the API you suggest) could contribute to the library.

PS: I am of course biased towards my own design, but I try to control that bias as best I can. Yet at the same time I have built very large form-based applications with forms spanning hundreds of controls with the current version of the library and found the design to be very successful in handling such scenarios.

**Comment by icepeng** [comment]
I like the current design, because it is more easy to read template code in my opinion.

```html
<form *ngIf="formState$ | async as formState" [ngrxFormState]="formState">
  <div *ngFor="let itemState of formState.controls.items.controls; trackBy: trackByIndex;">
    <input [ngrxFormControlState]="itemState" type="text">
  </div>
  <div *ngIf="formState.controls.items.isInvalid">
    <span *ngIf="formState.controls.items.errors.maxLength">Too many items</span>
  </div>
</form>
```

Looking to the code, it is very clear that itemState is control of form.items. And more easy to pass partial state into child component. If you write a template like Angular's `FormControlName`, there will be less boilerplate. But you have to track down all code to figure out the parent and ancestors of the form control.

And thanks for great library @MrWolfZ, I hope `ngrx-forms` to be merged into official ngrx platform.

**Comment by fxck** [comment]
look how "easy" is to find out whether form control is required or not with reactive forms..

https://github.com/angular/material2/issues/2574
https://github.com/angular/angular/issues/13461

**Comment by fxck** [comment]
There's this https://www.youtube.com/watch?v=kPkTXdToYV0 

It only shows how unnecessarily hard it is to work with forms. :/

**Comment by vincentjames501** [comment]
[3 principles from Redux](https://github.com/reduxjs/redux/blob/master/docs/introduction/ThreePrinciples.md). I personally think this should be the only argument necessary.

**Comment by brandonroberts** [comment]
Closing as non-actionable

**Comment by darpankumar** [comment]
@fxck @MrWolfZ can you provide your input on this discussion after 2 years now, still i could not see any solid reference or best practice to use

**Comment by fxck** [comment]
I've been using @MrWolfZ's lib for a while now, wrote a layer around it

![image](https://user-images.githubusercontent.com/1303561/89643462-1ff4b600-d8b6-11ea-9300-3c099b4e2f47.png)

it takes care of registering the form to a global slice of the ngrx state, has all the observable state of the form on it, along with helpers that return actions for you to dispatch, this part is basically a facade

![image](https://user-images.githubusercontent.com/1303561/89643558-49154680-d8b6-11ea-8b1e-4c4e5935bb77.png)

All in all, it works fine, has very similar to straightforward APIs that reactive forms have, but with the advantage of being able to change the state fully by dispatching actions and reading the state from ngrx store, it's little hacky inside, especially the main reducer, but works in all the cases I needed (including validations, focus management etc), never had the resources to make something community usable out of it though.

**Comment by darpankumar** [comment]
@fxck what u suggest to the community, is it ngrx/forms or any other idea which can help others also, or it would be great if u could write some article giving a brief idea about your work. as i could see u do have great knowledge related to this issue and u are the early bird of this discussion

**Comment by fxck** [comment]
It's hard, angular reactive forms suck, no matter if they make them typed or not, but ngrx-forms are extremely low level, too much for a widespread adoption, but given that ngrx team is not quite willing to take (any) forms project under their wings, I don't have a suggestion for you.

**Comment by EnricoVogt** [comment]
`angular reactive forms suck` just for my personal interest - can you please elaborate a little?

**Comment by fxck** [comment]
read the prior discussion to this issue 
https://github.com/ngrx/core/issues/12#issuecomment-273470234

amongst other things

- it's not truly reactive 
- it's nearly impossible to properly reset (submitted state is on ngForm directive)
- it's not type
- it doesn't play well with ngrx ecosystem (try reseting form from your effects)

**Comment by widiak** [comment]
We were also using ngrx-forms on project with around 2000 forms, and tables. Some of the forms were simple with 1-10 fields, some of them were "components" used as parts of bigger forms, some of  them were complex forms with lot of synchronous and asynchronous business logic, and we had also 2 complex 1000+ fields configurators. At the beginning we tried angular-forms, but it was too "closed" to fit our needs. We appreciated the "low-level" of ngrx-forms. 

We had to solve other problems, that ngrx-forms is not trying to solve (or was not trying to solve 2 years ago):
- Registering, and unregistering form-states into global state. It was a bit more complicated, because form identifier had to be calculated in runtime. 
- Cooperating ngrx-forms with ag-Grid, because a lot of tables were "editable" tables.
- We added a lot of custom metadata into state for some special form controls and because of use of ag-Grid.
- We did not like the @Effects - it was too simple, so we made separate layer for all asynchronous code and API to read and write to forms, and listen on form changes.
- We removed boxing/unboxing.
- We wanted making reducers to be more simple. So we have changed the ngrx-forms way (interface -> initial value -> update functions -> initial state -> reducer) to more simple syntax:
  ```ts
  this.form = this.formServiceFactory.create({
      id: ....,
      definition: defineGroup().addControls({
          userName: defineStringControl('User's login:', {
              update: validate(required, maxLength(40), minLength(5)),
          }),
          email: defineStringControl('User\'s email:', {
              create: setReadonly(true),
              update: validate(required),
          }),
          emailValidated: defineBooleanControl('Email validated', {
              defaultValue: false,
          })
      }, {
          update: state => {
              const emailReadonly = state.controls.userName.isValid === false;
              return setReadonly(emailReadonly, state, state.controls.email);
          },
      }),
   })
  ```
  Thus type of form is not resolved from interface, but from complex function, and so we had no problems with boxing/unboxing.
- Logic common to several controls is programmed in the update function of their common parent. The update function can only see and change the state of self and of its children.
- We are controlling reducer's "stability" - the reducer is executed several times and if the state  always change, we throw exception to developer.

What I am trying to say - we did a lot of complex things with ngrx-forms, but we were never blocked by ngrx-forms. It was always possible to modify ngrx-forms to suit our problems. This would be never possible with angular forms. And we are still happy with debugging the redux state in redux dev tools. The problems were with debuging huge states - so we had to do some optimalization so it is now usable, but copying big state from browser to redux dev tools is a bottle neck somtimes.

Also, at the end our modification of ngrx-forms was ok for junior programmers, because we have simplified the api for create/update state, and also because we created other layer to access form from "outside" reducer. We dont let application programmers to directly send messages into state.

So for me - ngrx-forms was perfect for complicated project, and I would appreceate if ngrx-forms would be a community project, so it can grow to perfect form library.

**Comment by Php-amqplibkieron2** [comment]
> It's hard, angular reactive forms suck, no matter if they make them typed or not, but ngrx-forms are extremely low level, too much for a widespread adoption. So given that ngrx team is not quite willing to take (any) forms project under their wings, I don't have a suggestion for you.





https://stackdev.space/questions/display-the-inputs-from-a-reactive-form-in-a-different-component

**Comment by widiak** [comment]
Regarding my previous comment from 2020:
It's 2025 now, and we still use ngrx-forms successfully (in our modified way) on other project with very (i mean veeery) complex main entity form.

**Comment by fxck** [comment]
Same.

**Comment by EnricoVogt** [comment]
What is the intention to bring ngrx-forms into ngrx-platform? Will it not maintained in the current repo by the maintainer?

**Comment by widiak** [comment]
I think there are 2 good reasons:

The main thing is that when it is "single maintainer repository", few projects would use it, because would fear the project will get frozen in time (single man show). So being part of ngrx means some "feeling" or "almost promise" of maintenance and continuous improvements. And it also helps marketing of the library (people simply consider useful to investigate what's new in ngrx community).

Other thing is that ngrx-forms could be better, maybe even much better, in how easily it can be used, extended, and also internally coded (typescript has improved too).  And this need some energy. And maybe it will be easier to find this energy for something covered with the "ngrx" mark, than for single-man repository.

**Comment by rainerhahnekamp** [comment]
> So being part of ngrx means some "feeling" or "almost promise" of maintenance

It is more of a promise or perceived guarantee that features are kept sharp. That's why it is important not to have too many features.

We also know that Angular's Signal Form will arrive soon. So it might be better to wait what they bring and don't take the risk of supporting features, which might become "outdated".

**Comment by widiak** [comment]
> It is more of a promise or perceived guarantee that features are kept sharp. That's why it is important not to have too many features.
> 
> We also know that Angular's Signal Form will arrive soon. So it might be better to wait what they bring and don't take the risk of supporting features, which might become "outdated".

I agree with both. Yet ...

The main thing is, that combination of ngrx and form-state gives developer a deterministic way of programming complex business and technical logic without surprises. I have not used Angular Signal in production, and not studied Signal-Form. Yet having enough experiences with Knockout observables in complex situation with not very senior team, I am in doubts. In ngrx you always know what change caused which consequences. In signals, i am not very sure about it. In ngrx (and specially ngrx-forms) you always know where to put the form's-code, when and why will be executed (because it is structured), and where you later find it. In signals, it depends on fantasy of developers how they connect signals together. Plus ngrx-forms gives you more freedom to store or manipulate state of form and offers possibilities of extending that state with more meta-informations, yet still in the same structure of code and still in predictable manner.

Btw: The first offer of ngrx-forms for ngrx platform came 8 years ago. In this discussion, 7 years ago was written: "I’m not sure the value add here over angular’s forms reactive forms implementation". I think reality of last 7 years with reactive forms proved, that it is really a time to dig deeply in ngrx-forms to understand it's treassure, and not waiting another 7 years if a new clone of old technique will be more successful than it's predecessor.

**Comment by fxck** [comment]
Big thing for me was that ngrx-forms include state that Angular's form do not, been so long I don't remember exactly what it was, something about form submitted state, which you couldn't change with FormControls, you had to ViewChild [ngForm] directive or something like that. It made managing form state from effects for example a nightmare. I'm not sure signal forms will solve that.

**Comment by rainerhahnekamp** [comment]
@widiak Wait a second. 

>  So it might be better to wait what they bring and don't take the risk of supporting features

That does not mean that once SignalForms are there, NgRx will support it.

First of all, I’m not the one to decide that. Second, SignalForms is a pre-requisite to even **think** about it. But again, this is a decision that the NgRx team has to make.

**Comment by widiak** [comment]
I think the first step is to deeply enough understand ngrx-forms library and then write a list of pros and cons (comparing to any other way of manipulating state of the form, either with signals or reactive forms). Then next step is to think about possibilities how ngrx-forms might look like in the future with reasonable effort (specially because the api is a bit technical and basically frozen how it was created in the past and some of us have experiences how to make it more user friendly). 

I hope this will gives us enough information to see the added value of combining ngrx/store and form-state in one library. Then it will be easeier to see if it make sense to put effort into such product. Other questions comes after, but we will at least know "why".
